### PR TITLE
ARROW-9861: [Java] Support big-endian in DecimalVector

### PR DESCRIPTION
**Pull request by kiszk**
This PR fixes failures in DecimalVectorTest on a big-endian platform

### CODE REVIEW

**Comment by emkornfield** [comment]
@kiszk you'll probably need to make some fixes for Decimal256 as well, I think.

@BryanCutler you mentioned you could devote some time to reviewing Big Endian changes?  Would you mind taking a look through this one and  @kiszk other Java changes?

**Comment by BryanCutler** [comment]
Sure, I can take a look. It might a day or two before I can though @kiszk .

**Comment by kiszk** [comment]
@emkornfield Sure, I will work for supporting Decimal256.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java [hunk 1]
```diff
@@ -151,15 +160,22 @@ private static void writeByteArrayToArrowBufHelper(byte[] bytes, ArrowBuf bytebu
...
-      bytesLE[i] = bytes[bytes.length - 1 - i];
+    if (LITTLE_ENDIAN) {
+      // Decimal stored as native-endian, need to swap data bytes before writing to ArrowBuf if LE
+      byte[] bytesLE = new byte[bytes.length];
+      for (int i = 0; i < bytes.length; i++) {
+        bytesLE[i] = bytes[bytes.length - 1 - i];
+      }
+
+      // Write LE data
+      byte [] padByes = bytes[0] < 0 ? minus_one : zeroes;
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:171]: you could move `padBytes` out of the `if` statements

File: java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java [hunk 2]
```diff
@@ -32,6 +32,31 @@
...
+      int [] intValues = new int [] {Integer.MAX_VALUE, Integer.MIN_VALUE, 0};
+      for (int val : intValues) {
+        buf.clear();
+        DecimalUtility.writeLongToArrowBuf((long) val, buf, 0);
+        BigDecimal actual = DecimalUtility.getBigDecimalFromArrowBuf(buf, 0, 0);
+        BigDecimal expected = BigDecimal.valueOf(val);
+        Assert.assertEquals(expected, actual);
+      }
+
+      long [] longValues = new long[] {Long.MIN_VALUE, 0 , Long.MAX_VALUE};
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java:49]: I don't think you need 2 loops here, just move the `Integer.MAX_VALUE` and `MIN_VALUE` here

**Review by BryanCutler — CHANGES_REQUESTED** [review]
LGTM, just a couple minor things

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java [hunk 3]
```diff
@@ -131,9 +135,14 @@ public static void writeBigDecimalToArrowBuf(BigDecimal value, ArrowBuf bytebuf,
    */
   public static void writeLongToArrowBuf(long value, ArrowBuf bytebuf, int index) {
     final long addressOfValue = bytebuf.memoryAddress() + (long) index * DECIMAL_BYTE_LENGTH;
-    PlatformDependent.putLong(addressOfValue, value);
     final long padValue = Long.signum(value) == -1 ? -1L : 0L;
-    PlatformDependent.putLong(addressOfValue + Long.BYTES, padValue);
+    if (LITTLE_ENDIAN) {
+      PlatformDependent.putLong(addressOfValue, value);
+      PlatformDependent.putLong(addressOfValue + Long.BYTES, padValue);
```
[the remark this answers is not in the document]
Comment (kiszk) [reply java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:141]: This is used for `DecialVector` that is a fixed-width (16-byte) vector. This routine extends the signed bit to a new long. I will write a document in the comment.

File: java/vector/src/main/java/org/apache/arrow/vector/DecimalVector.java [hunk 4]
```diff
@@ -218,25 +220,37 @@ public void setBigEndian(int index, byte[] value) {
...
+      if (length < TYPE_WIDTH) {
+        // sign extend
+        final byte pad = (byte) (value[0] < 0 ? 0xFF : 0x00);
+        PlatformDependent.setMemory(outAddress + length, DecimalVector.TYPE_WIDTH - length, pad);
+        return;
+      }
     } else {
-      throw new IllegalArgumentException(
-          "Invalid decimal value length. Valid length in [1 - 16], got " + length);
+      if (length <= TYPE_WIDTH) {
```
[the remark this answers is not in the document]
Comment (kiszk) [reply java/vector/src/main/java/org/apache/arrow/vector/DecimalVector.java:244]: If we add `if (length == TYPE_WIDTH) return` before line 244, no data is copied from `value` to `outAddress`. I will add a comment `copy data from value to outAddress` after line 244.

File: java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java (hunk 2 above)
Comment (kiszk) [reply java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java:49]: Sure, I will drop lines 49-57.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kiszk) [reply java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:141]: In addition, this PR will support to write it to a 256-bit entry for the future.

**Comment by kiszk** [comment]
I will update the Decimal256Vector class late today or tomorrow.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java [hunk 5]
```diff
@@ -128,12 +131,26 @@ public static void writeBigDecimalToArrowBuf(BigDecimal value, ArrowBuf bytebuf,
...
   /**
    * Write the given long to the ArrowBuf at the given value index.
+   * This routine extends the original sign bit to a new upper area in 128-bit or 256-bit.
    */
-  public static void writeLongToArrowBuf(long value, ArrowBuf bytebuf, int index) {
-    final long addressOfValue = bytebuf.memoryAddress() + (long) index * DECIMAL_BYTE_LENGTH;
-    PlatformDependent.putLong(addressOfValue, value);
+  public static void writeLongToArrowBuf(long value, ArrowBuf bytebuf, int index, int byteWidth) {
+    if (byteWidth != 16 && byteWidth != 32) {
+      throw new UnsupportedOperationException("DeciimalUtility.writeLongToArrowBuf() currently supports " +
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:138]: typo: `DeciimalUtility` -> `DecimalUtility`

File: java/vector/src/test/java/org/apache/arrow/vector/complex/impl/TestComplexCopier.java [hunk 6]
```diff
@@ -526,7 +526,7 @@ public void testCopyFixedSizedListOfDecimalsVector() {
       to.addOrGetVector(FieldType.nullable(new ArrowType.Decimal(3, 0, 128)));
 
       DecimalHolder holder = new DecimalHolder();
-      holder.buffer = allocator.buffer(DecimalUtility.DECIMAL_BYTE_LENGTH);
+      holder.buffer = allocator.buffer(16);
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/impl/TestComplexCopier.java:529]: Could you use `DecimalVector.TYPE_WIDTH` here?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 7]
```diff
@@ -310,7 +310,7 @@ public void listDecimalType() {
       listVector.allocateNew();
       UnionListWriter listWriter = new UnionListWriter(listVector);
       DecimalHolder holder = new DecimalHolder();
-      holder.buffer = allocator.buffer(DecimalUtility.DECIMAL_BYTE_LENGTH);
+      holder.buffer = allocator.buffer(16);
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java:313]: Same here

**Review by BryanCutler — COMMENTED** [review]
Looks good, just a few minor nits

**Review by BryanCutler — APPROVED** [review]
LGTM. Looks like a flaky test, will try again.

**Comment by BryanCutler** [comment]
The test error with Java JNI looks unrelated and seems to be an env issue with ORC, I'll go ahead with merging this.

**Comment by BryanCutler** [comment]
merged to master, thanks @kiszk !

**Comment by kiszk** [comment]
@BryanCutler Thank you. One comment.   
According to [the document](https://github.com/apache/arrow/blob/master/docs/source/developers/contributing.rst#endianness), benchmarking is necessary before merging features (I think that the test code does not affect the performance)
> Benchmarks for performance critical parts of the code to demonstrate no regression.

I am working for benchmark bot for Java [here](https://github.com/apache/arrow/pull/8210). It would be good to merge new features after this bot will be available.

cc @emkornfield

**Comment by BryanCutler** [comment]
I did not see anything that looks like it would affect performance here, but I agree we should get some benchmarks going to be sure. I will look at you other PR next.

**Pull request closed without merging**